api/extension-guides/language-model-tutorial.md

```diff
@@ -9,13 +9,325 @@ MetaDescription: Tutorial that walks you through creating a VS Code extension th
+  while (currentLine < endLine) {
+    code += `${currentLine + 1}: ${textEditor.document.lineAt(currentLine).text} \n`;
+    // move to the next line position
+    currentLine++;
+  }
+  return code;
+}
+```
+
+This code uses the `visibleRanges` property of the TextEditor to get the position of the lines that are currently visible in the editor. It then starts with the first line position and moves to the last line position, adding each line of code to a string along with the line number. Finally, it returns the string that contains all the viewable code with line numbers.
+
+Now we can call this method from the `code-tutor.annotate` command. Modify the implementation of the command so that it looks like this:
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getVisibleCodeWithLineNumbers(textEditor);
+
+});
+```
+
+### Step 2: Send code and prompt to language model API
+
+The next step is to call the GitHub Copilot language model and send it the user's code along with instructions to create the annotations.
+
+To do this, we first need to specify which chat model we want to use. We select 4o here because it is a fast and capable model for the kind of interaction we are building.
+
+```ts
+const disposable = vscode.commands.registerTextEditorCommand('code-tutor.annotate', async (textEditor: vscode.TextEditor) => {
+
+  // Get the code with line numbers from the current editor
+  const codeWithLineNumbers = getVisibleCodeWithLineNumbers(textEditor);
+
+  // select the 4o chat model
+  let [model] = await vscode.lm.selectChatModels({
+    vendor: 'copilot',
+    family: 'gpt-4o',
+  });
+});
+```
+
+We need instructions - or a "prompt" - that will tell the model to create the annotations and what format we want the response to be. Add the following code to the top of the file directly under the imports.
+
+```ts
+const ANNOTATION_PROMPT = `You are a code tutor who helps students learn how to write better code. Your job is to evaluate a block of code that the user gives you and then annotate any lines that could be improved with a brief suggestion and the reason why you are making that suggestion. Only make suggestions when you feel the severity is enough that it will impact the readibility and maintainability of the code. Be friendly with your suggestions and remember that these are students so they need gentle guidance. Format each suggestion as a single JSON object. It is not necessary to wrap your response in triple backticks. Here is an example of what your response should look like:
+
+{ "line": 1, "suggestion": "I think you should use a for loop instead of a while loop. A for loop is more concise and easier to read." }{ "line": 12, "suggestion": "I think you should use a for loop instead of a while loop. A for loop is more concise and easier to read." }
```

Comment:
To make the response easier to parse, you might ask for one JSON object per line, then you can check for a `\n` while parsing instead of a `}` which might appear in the text of a suggestion too.